diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt

```diff
@@ -0,0 +1,147 @@
+import com.pinterest.ktlint.core.ast.ElementType.KDOC
+import com.pinterest.ktlint.core.ast.ElementType.LATEINIT_KEYWORD
+import com.pinterest.ktlint.core.ast.ElementType.MODIFIER_LIST
+import com.pinterest.ktlint.core.ast.ElementType.OBJECT_DECLARATION
+import com.pinterest.ktlint.core.ast.ElementType.PRIVATE_KEYWORD
+import com.pinterest.ktlint.core.ast.ElementType.PROPERTY
+import com.pinterest.ktlint.core.ast.ElementType.REFERENCE_EXPRESSION
+import com.pinterest.ktlint.core.ast.ElementType.SECONDARY_CONSTRUCTOR
+import com.pinterest.ktlint.core.ast.ElementType.WHITE_SPACE
+import com.pinterest.ktlint.core.ast.nextSibling
+import com.pinterest.ktlint.core.ast.parent
+import com.pinterest.ktlint.core.ast.prevSibling
+import org.cqfn.diktat.common.config.rules.RulesConfig
+import org.cqfn.diktat.ruleset.constants.Warnings.BLANK_LINE_BETWEEN_PROPERTIES
+import org.cqfn.diktat.ruleset.constants.Warnings.WRONG_ORDER_IN_CLASS_LIKE_STRUCTURES
+import org.cqfn.diktat.ruleset.utils.checkOrder
+import org.cqfn.diktat.ruleset.utils.findAllNodesWithSpecificType
+import org.cqfn.diktat.ruleset.utils.findLeafWithSpecificType
+import org.cqfn.diktat.ruleset.utils.getIdentifierName
+import org.cqfn.diktat.ruleset.utils.leaveOnlyNNewLines
+import org.cqfn.diktat.ruleset.utils.moveChildBefore
+import org.jetbrains.kotlin.com.intellij.lang.ASTNode
+import org.jetbrains.kotlin.com.intellij.psi.tree.TokenSet
+
+/**
+ * Rule that checks order of declarations inside classes, interfaces and objects.
+ */
+class ClassLikeStructuresOrderRule : Rule("class-like-structures") {
+    private val loggerPropertyRegex = "(log|LOG|logger)".toRegex()
+
+    private lateinit var configRules: List<RulesConfig>
+    private lateinit var emitWarn: ((offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit)
+    private var isFixMode: Boolean = false
+
+    override fun visit(node: ASTNode,
+                       autoCorrect: Boolean,
+                       params: KtLint.Params,
+                       emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit) {
+        configRules = params.getDiktatConfigRules()
+        emitWarn = emit
+        isFixMode = autoCorrect
+
+        if (node.elementType == CLASS_BODY) {
+            checkDeclarationsOrderInClass(node)
+        } else if (node.elementType == PROPERTY) {
+            checkNewLinesBeforeProperty(node)
+        }
+    }
+
+    private fun checkDeclarationsOrderInClass(node: ASTNode) {
+        val allProperties = node.getChildren(TokenSet.create(PROPERTY))
+        val constProperties = allProperties.filter { it.findLeafWithSpecificType(CONST_KEYWORD) != null }
+        val lateInitProperties = allProperties.filter { it.findLeafWithSpecificType(LATEINIT_KEYWORD) != null }
+        val loggers = allProperties.filter {
+            it.findLeafWithSpecificType(PRIVATE_KEYWORD) != null && it.getIdentifierName()!!.text.contains(loggerPropertyRegex)
```

Comment:
loggers can be not only private, also they are commonly placed to companion object

Reply:
To quote our code style guide:
`Exception： All variants of a private val logger should be placed in the beginning of the class (private val log/LOG/logger/e.t.c)`.

It is quite vaguely stated. Should I change it both in code and in guide?

Reply:
ah, I got it! Please fix code style to: (private) val